File: XIV/Delphi/DynamicReverseProxy/ServerController.pas
unit ServerController;

interface

uses
  SysUtils, Classes, IWServerControllerBase, IWBaseForm, HTTPApp,
  // For OnNewSession Event
  UserSessionUnit, IWApplication, IWAppForm, IW.Browser.Browser,
  IW.HTTP.Request;

type
  TIWServerController = class(TIWServerControllerBase)
    procedure IWServerControllerBaseNewSession(ASession: TIWApplication);
    procedure IWServerControllerBaseRewriteURL(ARequest: THttpRequest; var ARewriteURL: string);

  private
    
  public
  end;


  function UserSession: TIWUserSession;
  function IWServerController: TIWServerController;

implementation

{$R *.dfm}

uses
  IWInit, IWGlobal;

function IWServerController: TIWServerController;
begin
  Result := TIWServerController(GServerController);
end;



function UserSession: TIWUserSession;
begin
  Result := TIWUserSession(WebApplication.Data);
end;

procedure TIWServerController.IWServerControllerBaseNewSession(
  ASession: TIWApplication);
begin
  ASession.Data := TIWUserSession.Create(nil, ASession);
end;


procedure TIWServerController.IWServerControllerBaseRewriteURL(ARequest: THttpRequest;
  var ARewriteURL: string);
var
  LBaseUrl: string;
begin
  LBaseUrl := ARequest.RawHeaders.Values['X-MyURLBase'];
  if LBaseUrl <> '' then
    ARewriteURL := LBaseUrl;
end;

initialization
  TIWServerController.SetServerControllerClass;

end.

File: XIV/C++/ReverseProxy/ServerController.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ServerController.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"

//---------------------------------------------------------------------------
__fastcall TIWServerController::TIWServerController(TComponent* Owner)
        : TIWServerControllerBase(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TIWServerController::IWServerControllerBaseNewSession(
      TIWApplication *ASession)
{
  ASession->Data = new TIWUserSession(NULL, ASession);
}
//---------------------------------------------------------------------------


void setServerController() {
  TIWServerController::SetServerControllerClass();
}

TIWServerController* IWServerController()
{
  return (TIWServerController*)gServerController;
}

#pragma startup setServerController
void __fastcall TIWServerController::IWServerControllerBaseRewriteURL(THttpRequest *ARequest,
		  UnicodeString &ARewriteURL)
{
  UnicodeString LBaseUrl = ARequest->GetRawHeaderValue("X-ProxyBase");
  if (!LBaseUrl.IsEmpty()) {
	ARewriteURL = LBaseUrl;
  }
}
//---------------------------------------------------------------------------

